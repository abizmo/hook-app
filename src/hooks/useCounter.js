import { useState } from "react"

const useCounter = (initialCounter = 0) => {
  const [counter, setCounter] = useState(initialCounter)

  const increment = ( factor = 1 ) => setCounter(counter => counter + factor)
  const decrement = ( factor = 1 ) => setCounter(counter => counter - factor)
  const reset = () => setCounter(initialCounter)

  return {
    counter,
    increment,
    decrement,
    reset,
  }
}

export default useCounter;
