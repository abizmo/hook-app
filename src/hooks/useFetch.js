import { useEffect, useRef, useState } from "react"

const useFetch = (url) => {
  const isMounted = useRef(true);
  const [{ data, error, loading }, setState] = useState({
    data: null,
    error: null,
    loading: true,
  })

  useEffect(() => {
    return () => {
      isMounted.current = false;
    }
  }, [])
  
  useEffect(() => {
    setState({
      data: null,
      error: null,
      loading: true,
    })
    fetch(url)
      .then(res => res.json())
      .then(data => isMounted.current && setState({
        data,
        error: null,
        loading: false,
      }))
      .catch(error => isMounted.current && setState({
        data: null,
        error,
        loading: false,
      }));
  }, [url])

  return { data, error, loading }
}

export default useFetch