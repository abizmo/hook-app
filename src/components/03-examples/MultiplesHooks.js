import React from 'react'

import useCounter from '../../hooks/useCounter'
import useFetch from '../../hooks/useFetch'

import './custom-hooks.css'

const MultiplesHooks = () => {
  const {counter, increment} = useCounter(1);
  const { data, loading, } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`)
  const { author, quote } = !!data && data[0]

  return (
    <div className="hooks-container">
      <h1>Breaking Bad Quotes</h1>
      <hr />
      {
        loading
        ? (
          <div className="alert alert-info text-center">
            Loading...
          </div>
        )
        : (
          <blockquote className="blockquote text-end">
            <p className="mb-2">{quote}</p>
            <footer className="blockquote-footer">{author}</footer>
          </blockquote>
        )
      }
      <button
        onClick={() => increment() }
        className="btn btn-primary"
      >
        Next
      </button>
    </div>
  )
}

export default MultiplesHooks
