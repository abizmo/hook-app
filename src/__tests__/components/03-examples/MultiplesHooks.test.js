import { shallow } from 'enzyme'
import MultiplesHooks from "../../../components/03-examples/MultiplesHooks"
import useCounter from '../../../hooks/useCounter'
import useFetch from '../../../hooks/useFetch'
jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe('MultiplesHooks test', () => {
  let useCounterResult;
  beforeEach(() => {
    useCounterResult = useCounter.mockReturnValue({
      counter: 1,
      increment: jest.fn()
    })
  })

  test('should render and match snapshot', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    })

    const wrapper = shallow(<MultiplesHooks />)

    expect(wrapper).toMatchSnapshot()
  })
  
  test('should show the info from the api', () => {
    useFetch.mockReturnValue({
      data: [{
        author: 'Fernando',
        quote: 'Hola Mundo',
      }],
      loading: false,
      error: null,
    })
    const wrapper = shallow(<MultiplesHooks />)

    expect(wrapper.find('.alert').exists()).toBeFalsy()
    expect(wrapper.find('p.mb-2').text().trim()).toBe('Hola Mundo')
    expect(wrapper.find('footer').text().trim()).toBe('Fernando')
  })
  
  test('should call increment when click on next button', () => {
    useFetch.mockReturnValue({
      data: [{
        author: 'Fernando',
        quote: 'Hola Mundo',
      }],
      loading: false,
      error: null,
    })
    const wrapper = shallow(<MultiplesHooks />)
    const nextBtn = wrapper.find('.btn')

    nextBtn.simulate('click')
    expect(useCounterResult).toHaveBeenCalled()
    expect(nextBtn.text().trim()).toBe('Next')
  })
})
