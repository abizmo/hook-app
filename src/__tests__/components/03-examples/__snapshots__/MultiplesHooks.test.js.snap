// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiplesHooks test should render and match snapshot 1`] = `
<div
  className="div hooks-container"
>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next
  </button>
</div>
`;
